modules/sr/robot3/arduino.py

```diff
@@ -0,0 +1,289 @@
+    """The analog pins on the Arduino."""
+    A0 = 14
+    A1 = 15
+    A2 = 16
+    A3 = 17
+    A4 = 18
+    A5 = 19
+
+    # Note: these pins don't exist on a real arduino.
+    A6 = 20
+    A7 = 21
+
+
+DIGITAL_READ_MODES = {GPIOPinMode.INPUT, GPIOPinMode.INPUT_PULLUP, GPIOPinMode.OUTPUT}
+DIGITAL_WRITE_MODES = {GPIOPinMode.OUTPUT}
+ANALOG_READ_MODES = {GPIOPinMode.INPUT}
+
+
+def init_arduinos(webot: Robot) -> MappingProxyType[str, Arduino]:
+    serial_number = '1234567890'
+    return MappingProxyType({
+        serial_number: Arduino(webot, serial_number),
+    })
+
+
+class Arduino:
+    """
+    The Arduino board interface.
+
+    This is intended to be used with Arduino Uno boards running the sbot firmware.
+
+    :param webot: The robot object to connect to devices on.
+    :param serial_number: The serial number of the board.
+    """
+    __slots__ = ('_serial_num', '_serial', '_pins', '_identity')
+
+    @staticmethod
+    def get_board_type() -> str:
+        """
+        Return the type of the board.
+
+        :return: The literal string 'Arduino'.
+        """
+        return 'Arduino'
+
+    def __init__(
+        self,
+        webot: Robot,
+        serial_number: str,
+    ) -> None:
+        # Stored for use in the identify method and repr.
+        self._serial_num = serial_number
+
+        self._pins = (
+            # Pins 0 and 1 are reserved for serial comms
+            Pin(0, supports_analog=False, disabled=True),
+            Pin(1, supports_analog=False, disabled=True),
+            # splat in both the digital and analog pins
+            *(Pin(index, supports_analog=False) for index in range(2, 14)),
+            *(Pin(index, supports_analog=True) for index in AnalogPins),
+        )
+
+        # Connect the devices to the pins.
+        # NOTE: the names here correspond to the names given to devices in Webots
+        # and, in some places, the keyboard controller.
+        self._pins[2]._connect_device(Microswitch(webot, 'back bump sensor'))
+        self._pins[3]._connect_device(Led(webot, 'led 1', pin_num=3))
+        self._pins[4]._connect_device(Led(webot, 'led 2', pin_num=4))
+        self._pins[AnalogPins.A0]._connect_device(DistanceSensor(webot, 'Front Left DS'))
+        self._pins[AnalogPins.A1]._connect_device(DistanceSensor(webot, 'Front Right DS'))
+        self._pins[AnalogPins.A2]._connect_device(DistanceSensor(webot, 'Left DS'))
+        self._pins[AnalogPins.A3]._connect_device(DistanceSensor(webot, 'Right DS'))
+        self._pins[AnalogPins.A4]._connect_device(DistanceSensor(webot, 'Front DS'))
+        self._pins[AnalogPins.A5]._connect_device(DistanceSensor(webot, 'Back DS'))
+        # NOTE: these pins don't exist on a real arduino.
+        self._pins[AnalogPins.A6]._connect_device(
+            PressureSensor(webot, 'finger pressure left'),
+        )
+        self._pins[AnalogPins.A7]._connect_device(
+            PressureSensor(webot, 'finger pressure right'),
+        )
```

Comment:
Configuration of which pins are doing what is not an inherent part of the Arduino and should probably be passed in as an argument.

Reply:
813d119db0f894315bfc2ace5ab8e45c69a0dc9d